repository/src/raw.rs

```diff
@@ -0,0 +1,189 @@
+    /// Deletes the branch.
+    async fn delete_branch(&mut self, branch: &Branch) -> Result<(), Error>;
+
+    // -------------------
+    // Tag-related methods
+    // -------------------
+
+    /// Returns the list of branches.
+    async fn list_tags(&self) -> Result<Vec<Tag>, Error>;
+
+    /// Creates a tag on the given commit.
+    async fn create_tag(&mut self, tag: &Tag, commit_hash: &CommitHash) -> Result<(), Error>;
+
+    /// Gets the commit that the tag points to.
+    async fn locate_tag(&self, tag: &Tag) -> Result<CommitHash, Error>;
+
+    /// Gets the tags on the given commit.
+    async fn get_tag(&self, commit_hash: &CommitHash) -> Result<Vec<Tag>, Error>;
+
+    /// Removes the tag.
+    async fn remove_tag(&mut self, tag: &Tag) -> Result<(), Error>;
+
+    // ----------------------
+    // Commit-related methods
+    // ----------------------
+
+    /// Create a commit with the branch.
+    ///
+    /// Same as `checkout_clean() && git checkout <branch> && git add -A && git commit`.
+    async fn create_commit(
+        &mut self,
+        branch: &Branch,
+        commit_message: &str,
+        diff: Option<&str>,
+    ) -> Result<(), Error>;
+
+    /// Create a commit with the given list to overwrite files.
+    ///
+    /// * `files`: the list of `(file_path, file_content_to_overwrite)`.
+    async fn create_commit_overwrite(
+        &mut self,
+        branch: &Branch,
+        commit_message: &str,
+        files: Vec<(String, String)>,
+    ) -> Result<(), Error>;
+
+    /// Removes orphaned commits. Same as `git gc --prune=now --aggressive`
+    async fn run_garbage_collection(&mut self) -> Result<(), Error>;
+
+    // ----------------------------
+    // Working-tree-related methods
+    // ----------------------------
+
+    /// Checkouts and cleans the current working tree.
+    /// This is same as `git checkout . && git clean -fd`.
+    async fn checkout_clean(&mut self) -> Result<(), Error>;
+
+    /// Checkouts to the branch.
+    async fn checkout(&mut self, branch: &Branch) -> Result<(), Error>;
+
+    /// Checkouts to the commit and make `HEAD` in a detached mode.
+    async fn checkout_detach(&mut self, commit_hash: &CommitHash) -> Result<(), Error>;
+
+    // ---------------
+    // Various queries
+    // ---------------
+
+    /// Returns the commit hash of the current HEAD.
+    async fn get_head(&self) -> Result<CommitHash, Error>;
+
+    /// Returns the commit hash of the initial commit.
+    ///
+    /// Fails if the repository is empty.
+    async fn get_initial_commit(&self) -> Result<CommitHash, Error>;
+
+    /// Returns the diff of the given commit.
+    async fn show_commit(&self, commit_hash: &CommitHash) -> Result<String, Error>;
+
+    /// Lists the ancestor commits of the given commit (The first element is the direct parent).
+    ///
+    /// It fails if there is a merge commit.
+    async fn list_ancestors(
+        &self,
+        commit_hash: &CommitHash,
+        max: Option<usize>,
```

Comment:
Is "max" used to restrict the number of ancestor commits?

Reply:
comment added.